File: clients/csharp/Ttid.cs
// TTID client — drives the `ttid` binary's persistent NDJSON loop.
//
// No NuGet dependencies (System.Text.Json ships with .NET). Requires the `ttid`
// binary on PATH or an explicit path. One long-lived subprocess.
//
//   using var t = new Ttid();
//   string id = t.Generate().GetString();          // new id
//   JsonElement up = t.Generate(id);               // advance it
//   JsonElement times = t.DecodeTime(up.GetString());
//   JsonElement valid = t.IsTTID(id);
//   JsonElement uuid = t.IsUUID("...");
//
// Each method builds the request and returns the op's `result` as a JsonElement
// (throwing TtidException on failure). Method names follow .NET PascalCase.
// Request(json) is the raw escape hatch returning the full response.

using System;
using System.Diagnostics;
using System.Text.Json;

namespace Ttid
{
    public sealed class TtidException : Exception
    {
        public TtidException(string message) : base(message) { }
    }

    public sealed class Ttid : IDisposable
    {
        private readonly Process _proc;
        private readonly object _lock = new object();

        public Ttid(string binary = "ttid")
        {
            var psi = new ProcessStartInfo
            {
                FileName = binary,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                UseShellExecute = false,
            };
            psi.ArgumentList.Add("exec");
            psi.ArgumentList.Add("--loop");
            _proc = Process.Start(psi) ?? throw new InvalidOperationException("failed to start ttid");
        }

        /// <summary>Send one raw machine-protocol op (JSON string); returns the full response.</summary>
        public JsonDocument Request(string opJson)
        {
            lock (_lock) // ponytail: one call in flight; drop the lock only if you pipeline
            {
                if (_proc.HasExited) throw new InvalidOperationException("ttid process has exited");
                _proc.StandardInput.Write(opJson.TrimEnd());
                _proc.StandardInput.Write('\n');
                _proc.StandardInput.Flush();
                string? line = _proc.StandardOutput.ReadLine();
                if (line == null) throw new InvalidOperationException("ttid closed the stream");
                return JsonDocument.Parse(line);
            }
        }

        // Send a fully-formed op JSON and return `result`, throwing on failure.
        private JsonElement Op(string opJson)
        {
            using JsonDocument doc = Request(opJson);
            JsonElement root = doc.RootElement;
            if (!root.GetProperty("ok").GetBoolean())
            {
                string msg = root.TryGetProperty("error", out var e) &&
                             e.TryGetProperty("message", out var m)
                    ? m.GetString() ?? "ttid error"
                    : "ttid error";
                throw new TtidException(msg);
            }
            return root.TryGetProperty("result", out var r) ? r.Clone() : default;
        }

        private static string J(string value) => JsonSerializer.Serialize(value);

        /// <summary>Generate a new TTID, or advance <paramref name="id"/> (del=true to tombstone).</summary>
        public JsonElement Generate(string? id = null, bool del = false)
        {
            string op = "{\"op\":\"generate\"";
            if (id != null) op += $",\"id\":{J(id)}";
            if (del) op += ",\"delete\":true";
            return Op(op + "}");
        }

        public JsonElement DecodeTime(string id) =>
            Op($"{{\"op\":\"decodeTime\",\"id\":{J(id)}}}");

        public JsonElement IsTTID(string id) =>
            Op($"{{\"op\":\"isTTID\",\"id\":{J(id)}}}");

        public JsonElement IsUUID(string id) =>
            Op($"{{\"op\":\"isUUID\",\"id\":{J(id)}}}");

        public void Dispose()
        {
            if (!_proc.HasExited)
            {
                _proc.StandardInput.Close(); // EOF ends the loop
                _proc.WaitForExit(30_000);
            }
            _proc.Dispose();
        }
    }
}

File: clients/rust/ttid.rs
//! TTID client — drives the `ttid` binary's persistent NDJSON loop.
//!
//! std only. Requires the `ttid` binary on PATH or an explicit path. One
//! long-lived subprocess. Methods build the request and return the raw response
//! line (also JSON); bring serde if you want typed structs.
//!
//!   let mut t = Ttid::open("ttid")?;
//!   let id = t.generate(None, false)?;                  // {"...","result":"4VL..."}
//!   let up = t.generate(Some("4VL..."), false)?;        // advance it
//!   let times = t.decode_time("4VL...")?;
//!   let valid = t.is_ttid("4VL...")?;
//!   let uuid = t.is_uuid("...")?;
//!
//! Method names follow Rust's snake_case; `request` is the raw escape hatch for
//! ops without a method.

use std::io::{BufRead, BufReader, Write};
use std::process::{Child, ChildStdin, Command, Stdio};

pub struct Ttid {
    child: Child,
    stdin: Option<ChildStdin>,
    stdout: BufReader<std::process::ChildStdout>,
}

impl Ttid {
    /// Start a warm ttid process. `binary` is usually "ttid".
    pub fn open(binary: &str) -> std::io::Result<Ttid> {
        let mut child = Command::new(binary)
            .args(["exec", "--loop"])
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()?;
        let stdin = child.stdin.take().expect("stdin piped");
        let stdout = BufReader::new(child.stdout.take().expect("stdout piped"));
        Ok(Ttid { child, stdin: Some(stdin), stdout })
    }

    /// Send one machine-protocol op (a JSON object string) and return the response line.
    pub fn request(&mut self, op_json: &str) -> std::io::Result<String> {
        let stdin = self
            .stdin
            .as_mut()
            .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::BrokenPipe, "ttid closed"))?;
        stdin.write_all(op_json.trim_end().as_bytes())?;
        stdin.write_all(b"\n")?;
        stdin.flush()?;
        let mut line = String::new();
        if self.stdout.read_line(&mut line)? == 0 {
            return Err(std::io::Error::new(
                std::io::ErrorKind::UnexpectedEof,
                "ttid closed the stream",
            ));
        }
        Ok(line)
    }

    // Send a fully-formed op JSON and error on a failure response.
    // ponytail: checks the always-present "ok":true field by substring.
    fn checked(&mut self, json: String) -> std::io::Result<String> {
        let resp = self.request(&json)?;
        if !resp.contains("\"ok\":true") {
            return Err(std::io::Error::new(std::io::ErrorKind::Other, resp.trim().to_string()));
        }
        Ok(resp)
    }

    /// Generate a new TTID, or advance `id` (`del` to tombstone). `None` for a fresh id.
    pub fn generate(&mut self, id: Option<&str>, del: bool) -> std::io::Result<String> {
        let mut op = String::from(r#"{"op":"generate""#);
        if let Some(id) = id {
            op.push_str(&format!(r#","id":"{}""#, esc(id)));
        }
        if del {
            op.push_str(r#","delete":true"#);
        }
        op.push('}');
        self.checked(op)
    }

    pub fn decode_time(&mut self, id: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"decodeTime","id":"{}"}}"#, esc(id)))
    }

    pub fn is_ttid(&mut self, id: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"isTTID","id":"{}"}}"#, esc(id)))
    }

    pub fn is_uuid(&mut self, id: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"isUUID","id":"{}"}}"#, esc(id)))
    }

    /// End the loop and wait for the process to exit.
    pub fn close(mut self) -> std::io::Result<()> {
        self.stdin.take(); // drop stdin → EOF ends the loop
        self.child.wait().map(|_| ())
    }
}

/// Escape a string for embedding in a JSON string literal.
fn esc(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out
}
